castor-service/src/main/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingService.java

```diff
@@ -117,26 +127,64 @@ public void keepReservation(Reservation reservation) {
    * @throws CastorServiceException if the tuples could not be reserved as requested.
    * @throws CastorServiceException if the cache already holds a reservation with the given ID
    */
-  public void keepAndApplyReservation(Reservation reservation) {
+  @Transactional
+  public void keepAndApplyReservation(Reservation reservation) throws InterruptedException {
     log.debug("persisting reservation {}", reservation);
     ValueOperations<String, Object> ops = redisTemplate.opsForValue();
     if (ops.get(cachePrefix + reservation.getReservationId()) == null) {
-      ops.set(cachePrefix + reservation.getReservationId(), reservation);
-      log.debug("put in database at {}", cachePrefix + reservation.getReservationId());
+      //      TransactionSynchronizationManager.registerSynchronization(new
+      // TransactionSynchronization() {
+      //        @Override
+      //        public void afterCommit(){
+      //          log.debug("put in database at {}", cachePrefix + reservation.getReservationId());
+      //
+      //        }
+      //      });
       log.debug("Apply reservation {}", reservation);
-      for (ReservationElement re : reservation.getReservations()) {
-        log.debug("Processing reservation element {}", re);
-        long startIndex = re.getStartIndex();
-        long endIndex = startIndex + re.getReservedTuples();
-        while (startIndex < endIndex) {
-          TupleChunkFragmentEntity fragment =
-              tupleChunkFragmentStorageService
-                  .findAvailableFragmentForChunkContainingIndex(re.getTupleChunkId(), startIndex)
-                  .orElseThrow(
-                      () ->
-                          new CastorServiceException(
-                              String.format(
-                                  RESERVATION_CANNOT_BE_SATISFIED_EXCEPTION_FORMAT, reservation)));
+      storeReservationInDB(reservation);
+      ops.set(cachePrefix + reservation.getReservationId(), reservation);
+      applyConsumption(reservation);
+      log.debug("consumption emitted");
+    } else {
+      System.err.println("reservation conflict");
+      throw new CastorServiceException(
+          String.format(RESERVATION_CONFLICT_EXCEPTION_MSG, reservation.getReservationId()));
+    }
+  }
+
+  /**
+   * @param reservation The reserved fragments
+   * @throws InterruptedException : If the fragments can not be reserved
+   */
+  @Transactional()
+  public void storeReservationInDB(@NotNull Reservation reservation) throws InterruptedException {
```

Comment:
Am I right that this method requires the reservation elements to be ordered by number of reserved Tuples where all elements with number of tuples < initial fragment size must be listed first.
That feels fragile especially as not properly documented (is it?)